File: backend/templates/fibonacci/fibonacci.php
<?php

// 1. 再帰版（メモ化付き）
class FibonacciRecursive {
    private $memo = [];
    
    public function calculate($n) {
        if ($n <= 1) return $n;
        if (isset($this->memo[$n])) {
            return $this->memo[$n];
        }
        $this->memo[$n] = $this->calculate($n - 1) + $this->calculate($n - 2);
        return $this->memo[$n];
    }
}

// 2. イテレータ版
class FibonacciIterator implements Iterator {
    private $limit;
    private $position = 0;
    private $a = 0;
    private $b = 1;
    private $current = 0;
    
    public function __construct($limit = PHP_INT_MAX) {
        $this->limit = $limit;
    }
    
    public function rewind(): void {
        $this->position = 0;
        $this->a = 0;
        $this->b = 1;
        $this->current = 0;
    }
    
    public function current() {
        return $this->current;
    }
    
    public function key() {
        return $this->position;
    }
    
    public function next(): void {
        if ($this->position == 0) {
            $this->current = $this->a;
        } else {
            $temp = $this->a;
            $this->a = $this->b;
            $this->b = $temp + $this->b;
            $this->current = $temp;
        }
        $this->position++;
    }
    
    public function valid(): bool {
        return $this->position < $this->limit;
    }
    
    public function toArray() {
        $result = [];
        foreach ($this as $value) {
            $result[] = $value;
        }
        return $result;
    }
}

// 3. 行列累乗版
class FibonacciMatrix {
    public function calculate($n) {
        if ($n <= 1) return $n;
        $matrix = [[1, 1], [1, 0]];
        $result = $this->matrixPower($matrix, $n);
        return $result[0][1];
    }
    
    private function matrixMultiply($a, $b) {
        return [
            [$a[0][0] * $b[0][0] + $a[0][1] * $b[1][0], $a[0][0] * $b[0][1] + $a[0][1] * $b[1][1]],
            [$a[1][0] * $b[0][0] + $a[1][1] * $b[1][0], $a[1][0] * $b[0][1] + $a[1][1] * $b[1][1]]
        ];
    }
    
    private function matrixPower($matrix, $n) {
        if ($n == 0) return [[1, 0], [0, 1]];
        if ($n == 1) return $matrix;
        
        if ($n % 2 == 0) {
            $half = $this->matrixPower($matrix, intval($n / 2));
            return $this->matrixMultiply($half, $half);
        } else {
            return $this->matrixMultiply($matrix, $this->matrixPower($matrix, $n - 1));
        }
    }
}

// 4. ジェネレータ版
function fibonacciGenerator() {
    $a = 0;
    $b = 1;
    while (true) {
        yield $a;
        $temp = $a;
        $a = $b;
        $b = $temp + $b;
    }
}

// 5. 黄金比を使った近似計算
class FibonacciGoldenRatio {
    const PHI = 1.6180339887498948;
    
    public function calculate($n) {
        return round((pow(self::PHI, $n) - pow(-self::PHI, -$n)) / sqrt(5));
    }
}

// 6. 配列を使った版
function fibonacciArray($n) {
    if ($n <= 0) return [];
    if ($n == 1) return [0];
    
    $fib = [0, 1];
    for ($i = 2; $i < $n; $i++) {
        $fib[$i] = $fib[$i-1] + $fib[$i-2];
    }
    return $fib;
}

// デモンストレーション
echo "=== フィボナッチ数列の様々な実装 ===\n\n";

// 1. 再帰版のデモ
echo "1. 再帰版（メモ化付き）\n";
$fibRecursive = new FibonacciRecursive();
echo "最初の10項: ";
for ($i = 0; $i < 10; $i++) {
    echo $fibRecursive->calculate($i) . " ";
}
echo "\n\n";

// 2. イテレータ版のデモ
echo "2. イテレータ版\n";
echo "最初の10項: ";
$fibIter = new FibonacciIterator(10);
foreach ($fibIter as $value) {
    echo $value . " ";
}
echo "\n\n";

// 3. 行列累乗版のデモ
echo "3. 行列累乗版（大きな数も高速）\n";
$fibMatrix = new FibonacciMatrix();
echo "n=50の値: " . $fibMatrix->calculate(50) . "\n\n";

// 4. ジェネレータ版のデモ
echo "4. ジェネレータ版（遅延評価）\n";
echo "100未満の項: ";
foreach (fibonacciGenerator() as $value) {
    if ($value >= 100) break;
    echo $value . " ";
}
echo "\n\n";

// 5. 黄金比版のデモ
echo "5. 黄金比を使った近似計算\n";
$fibGolden = new FibonacciGoldenRatio();
echo "最初の10項: ";
for ($i = 0; $i < 10; $i++) {
    echo $fibGolden->calculate($i) . " ";
}
echo "\n\n";

// 6. 配列版のデモ
echo "6. 配列版\n";
$fibArray = fibonacciArray(15);
echo "最初の15項: " . implode(" ", $fibArray) . "\n\n";

// おまけ：フィボナッチ数列の性質を調べる
echo "=== フィボナッチ数列の面白い性質 ===\n";

// 隣接項の比が黄金比に収束
echo "隣接項の比（黄金比" . number_format(FibonacciGoldenRatio::PHI, 6) . "に収束）:\n";
for ($i = 5; $i <= 10; $i++) {
    $ratio = $fibArray[$i] / $fibArray[$i-1];
    echo "F($i)/F(" . ($i-1) . ") = " . number_format($ratio, 6) . "\n";
}

// 最初のn項の和
echo "\n最初のn項の和の性質（F(n+2) - 1に等しい）:\n";
for ($n = 5; $n <= 8; $n++) {
    $sum = array_sum(array_slice($fibArray, 0, $n));
    echo "最初の{$n}項の和: $sum = F(" . ($n+2) . ") - 1 = " . $fibArray[$n+1] . " - 1\n";
}

// PHPの特殊機能を使った実装
echo "\n=== PHPの特殊機能を使った実装 ===\n";

// クロージャ版
$fibonacciClosure = function($n) use (&$fibonacciClosure) {
    static $memo = [];
    if ($n <= 1) return $n;
    if (isset($memo[$n])) return $memo[$n];
    return $memo[$n] = $fibonacciClosure($n-1) + $fibonacciClosure($n-2);
};

echo "クロージャ版 F(10): " . $fibonacciClosure(10) . "\n";

// 匿名関数とarray_mapを使った版
$numbers = range(0, 9);
$fibValues = array_map($fibonacciClosure, $numbers);
echo "匿名関数版（最初の10項）: " . implode(" ", $fibValues) . "\n";

?>

File: backend/templates/fibonacci/fibonacci.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <cmath>
#include <iomanip>

// 1. 再帰版（メモ化付き）
class FibonacciRecursive {
private:
    std::unordered_map<int, long long> memo;

public:
    long long calculate(int n) {
        if (n <= 1) return n;
        if (memo.find(n) != memo.end()) {
            return memo[n];
        }
        memo[n] = calculate(n - 1) + calculate(n - 2);
        return memo[n];
    }
};

// 2. イテレータ版
class FibonacciIterator {
private:
    int limit;
    int current;
    long long a, b;

public:
    FibonacciIterator(int limit) : limit(limit), current(0), a(0), b(1) {}
    
    bool hasNext() const {
        return current < limit;
    }
    
    long long next() {
        if (current == 0) {
            current++;
            return a;
        }
        long long temp = a;
        a = b;
        b = temp + b;
        current++;
        return temp;
    }
    
    std::vector<long long> toVector() {
        std::vector<long long> result;
        FibonacciIterator iter(limit);
        while (iter.hasNext()) {
            result.push_back(iter.next());
        }
        return result;
    }
};

// 3. 行列累乗版
class Matrix {
public:
    long long m[2][2];
    
    Matrix() {
        m[0][0] = m[0][1] = m[1][0] = m[1][1] = 0;
    }
    
    Matrix(long long a, long long b, long long c, long long d) {
        m[0][0] = a; m[0][1] = b;
        m[1][0] = c; m[1][1] = d;
    }
    
    Matrix operator*(const Matrix& other) const {
        Matrix result;
        result.m[0][0] = m[0][0] * other.m[0][0] + m[0][1] * other.m[1][0];
        result.m[0][1] = m[0][0] * other.m[0][1] + m[0][1] * other.m[1][1];
        result.m[1][0] = m[1][0] * other.m[0][0] + m[1][1] * other.m[1][0];
        result.m[1][1] = m[1][0] * other.m[0][1] + m[1][1] * other.m[1][1];
        return result;
    }
};

class FibonacciMatrix {
public:
    long long calculate(int n) {
        if (n <= 1) return n;
        Matrix base(1, 1, 1, 0);
        Matrix result = matrixPower(base, n);
        return result.m[0][1];
    }

private:
    Matrix matrixPower(Matrix base, int n) {
        if (n == 0) return Matrix(1, 0, 0, 1); // 単位行列
        if (n == 1) return base;
        
        if (n % 2 == 0) {
            Matrix half = matrixPower(base, n / 2);
            return half * half;
        } else {
            return base * matrixPower(base, n - 1);
        }
    }
};

// 4. 黄金比を使った近似計算
class FibonacciGoldenRatio {
private:
    static constexpr double PHI = (1.0 + std::sqrt(5.0)) / 2.0;

public:
    long long calculate(int n) {
        return static_cast<long long>(std::round(
            (std::pow(PHI, n) - std::pow(-PHI, -n)) / std::sqrt(5.0)
        ));
    }
    
    static double getPhi() { return PHI; }
};

// 5. テンプレートを使ったジェネリック版
template<typename T>
class FibonacciGeneric {
public:
    static std::vector<T> generate(int n) {
        std::vector<T> result;
        if (n >= 1) result.push_back(0);
        if (n >= 2) result.push_back(1);
        for (int i = 2; i < n; i++) {
            result.push_back(result[i-1] + result[i-2]);
        }
        return result;
    }
};

int main() {
    std::cout << "=== フィボナッチ数列の様々な実装 ===" << std::endl << std::endl;
    
    // 1. 再帰版のデモ
    std::cout << "1. 再帰版（メモ化付き）" << std::endl;
    FibonacciRecursive fibRecursive;
    std::cout << "最初の10項: ";
    for (int i = 0; i < 10; i++) {
        std::cout << fibRecursive.calculate(i) << " ";
    }
    std::cout << std::endl << std::endl;
    
    // 2. イテレータ版のデモ
    std::cout << "2. イテレータ版" << std::endl;
    std::cout << "最初の10項: ";
    FibonacciIterator fibIter(10);
    while (fibIter.hasNext()) {
        std::cout << fibIter.next() << " ";
    }
    std::cout << std::endl << std::endl;
    
    // 3. 行列累乗版のデモ
    std::cout << "3. 行列累乗版（大きな数も高速）" << std::endl;
    FibonacciMatrix fibMatrix;
    std::cout << "n=50の値: " << fibMatrix.calculate(50) << std::endl << std::endl;
    
    // 4. 黄金比版のデモ
    std::cout << "4. 黄金比を使った近似計算" << std::endl;
    FibonacciGoldenRatio fibGolden;
    std::cout << "最初の10項: ";
    for (int i = 0; i < 10; i++) {
        std::cout << fibGolden.calculate(i) << " ";
    }
    std::cout << std::endl << std::endl;
    
    // 5. テンプレート版のデモ
    std::cout << "5. テンプレート版（ジェネリック）" << std::endl;
    auto sequence = FibonacciGeneric<long long>::generate(15);
    std::cout << "最初の15項: ";
    for (const auto& num : sequence) {
        std::cout << num << " ";
    }
    std::cout << std::endl << std::endl;
    
    // おまけ：フィボナッチ数列の性質を調べる
    std::cout << "=== フィボナッチ数列の面白い性質 ===" << std::endl;
    
    // 隣接項の比が黄金比に収束
    std::cout << std::fixed << std::setprecision(6);
    std::cout << "隣接項の比（黄金比" << FibonacciGoldenRatio::getPhi() << "に収束）:" << std::endl;
    for (int i = 5; i <= 10; i++) {
        double ratio = static_cast<double>(sequence[i]) / sequence[i-1];
        std::cout << "F(" << i << ")/F(" << i-1 << ") = " << ratio << std::endl;
    }
    
    // 最初のn項の和
    std::cout << std::endl << "最初のn項の和の性質（F(n+2) - 1に等しい）:" << std::endl;
    for (int n = 5; n <= 8; n++) {
        long long sum = 0;
        for (int i = 0; i < n; i++) {
            sum += sequence[i];
        }
        std::cout << "最初の" << n << "項の和: " << sum 
                  << " = F(" << n+2 << ") - 1 = " << sequence[n+1] << " - 1" << std::endl;
    }
    
    return 0;
}